### ISSUE TITLE
first error on electrum: key

### ISSUE THREAD

**Issue by venzen** [spesmilo/electrum#1919]
What could this mean?:

File "/home/venzen/electrum/lib/wallet.py", line 477, in get_addr_utxo
    coins.pop(txi)
KeyError: u'_obfuscated_:1'

I obfuscated the full 64 char length hash.

This traceback displays at startup and any attempt to send funds pops up "Invalid Fee" regardless of fee settings

**Comment by venzen** [comment]
Anyone? I'm keeping a cool heart even though this was a big fucking disaster for me with Electrum today. I'd be interested in understanding why Electrum suddenly throws this error when receiving a payment when it worked as expected for years through various upgrades.

**Comment by dabura667** [comment]
Do you have your seed written down?

Try restoring your seed into a separate wallet file (by clicking "New Wallet" and restoring, give it a different filename than your original.)

I know what the error is, and this will probably get around it temporarily, but as for the cause of the error, I haven't kept up with Electrum source enough, nor do I have much time to look into it, sorry.

**Comment by ecdsa** [comment]
looks like some wallet corruption.
impossible to tell more unless you share the content of the wallet file.

as @dabura667 said, restoring from seed should work

**Comment by venzen** [comment]
Thanks for the feedback. This happens to be a non-seed wallet to which I imported keys from a collection of other older wallets. So no seed. I accessed the funds by exporting the relevant key to another wallet.

Does this pattern of corruption conform to a known internal error?

**Comment by venzen** [comment]
@dabura667, perhaps you can give me a brief synopsis and a pointer or two. I'm on a high octane diet of coffee and sugar cane and can probably fix it with my jungle Python-fu.

**Comment by ecdsa** [comment]
no it does not correspond to a known error.

**Comment by ecdsa** [comment]
if you want me to have a look, please remove the seed and private keys from the wallet file, and email it to me.

**Comment by dabura667** [comment]
@venzen Basically, the code is deciding which utxos to use in generating the transaction, and uses pop on the coins dict... the items of this dict have keys that correspond to their txid and the utxo index. (So yours says <txid>:1 so it means "I want to pop the item in coins with the key of <txid>:1 which is the 2nd output of the tx <txid>")

What is probably happening:

Your wallet file has the same utxo:index set written twice somewhere. (This would make sense as if you add two things with the same key in a dict in python, it will over write the item, but when you pop it, it can only pop once, which the second attempt would give a KeyError)

In that case, I would probably go into the wallet and try to export the private keys. If you can't figure that out, I would recommend opening the wallet file in some sort of text editor like Atom or Notepad++ etc. and look for the area with your private keys.

It should look something like: `MSUVK70k6uK04TwUqxe6QccyhjxwR2RXuyipKaI+9Vk=` but much longer.

Then go into Electrum and in the console tab, type:

# WARNING: BEFORE TYPING THIS IN, KNOW THAT YOUR PASSWORD WILL BE SAVED TO YOUR WALLET FILE ON DISK. Because Electrum saves a history of all commands typed into the console tab to your wallet file.

```python
>> from electrum.bitcoin import pw_decode
>> pw_decode("MSUVK70k6uK04TwUqxe6QccyhjxwR2RXuyipKaI+9Vk=", "test2")
u'test'
```

where test2 is your password for encryption and the MSU...k= string is the encrypted private keys. It will show you all your private keys so you can re-import them.

There is a line in the https://github.com/spesmilo/electrum/blob/master/gui/qt/console.py#L147 which says that adding a space to the beginning of the command will not record in in history, but when I tried it, I got an indentation error from the python interpreter... so in the absence of a better way, I would just go into the wallet file afterward and delete the list item manually that contained your password...

**Comment by sup2X** [comment]
I get the same error irregularly when I'm sending bitcoins with RPC API. When the error appears I cannot send the bitcoins until I restart electrum daemon.

`File "/usr/local/lib/python2.7/dist-packages/electrum/wallet.py", line 612, in get_addr_utxo | KeyError: u'3 ... bd:1'`

It happens with two different wallets. Recovering from the seed didn't help to resolve the problem permanently.

Can it be related to pruning limits? The address u'3 ... bd:1' has more than 200 transactions.

**Comment by venzen** [comment]
@ecdsa , @dabura667 

I already had backups of all the private keys that had been imported to this wallet. So, I simply created another wallet, imported them again and the error has not occurred again. 

When I originally did this, I must have imported the offending key twice...

**Comment by venzen** [comment]
OK, shortly after posting my previous comment the original error returned to the newly created wallet. The offending transaction id (obfuscated in the console error output in my first post) involved a key that had, in previous non-Electrum wallets, been used as a change address. 

Similar to the case mentioned by @sup2X it had been involved in over a hundred txns. I manually swept its total value to my Electrum seed wallet. As soon as that sweep transaction confirmed the console errors stopped. 

The implication is that the wallet ```KeyError``` has to do either with: 

1. the number of UTXOs associated with this address, or 
2. the absence of an associated address/privkey used as a change address somewhere along the way in this address's rolling stone life in another wallet

Sweeping the private key solved the issue.

I've searched for an answer but haven't found: How do I delete a privkey from Electrum?

**Comment by ecdsa** [comment]
does the address have more than 1000 utxos?

**Comment by venzen** [comment]
I don't believe it has. Blockchain.info reports 112 txns, as does Electrum. But I'm unsure how to get at the utxos...

getaddressunspent returns an empty list for this transaction. Is there another command that will return utxo count?

**Comment by venzen** [comment]
it has 999 utxos. When someone else hollers we'll deal with it then ok?

**Comment by ecdsa** [comment]
reopening. the crash should still be addressed.

**Comment by venzen** [comment]
is there a way i can help, with logs or an empty wallet file?

If anyone can assist me to list the utxos (either via electrum or bitcoind) I can provide feedback about that.

**Comment by sup2X** [comment]
I think that it happens only if an address in the wallet has more than 100 transactions and when electrum switches from server with lower pruning limit (100) to server with higher pruning limit (10000), or in the other way. It could create inconsistency in utxo list.

But I'm just guessing (I didn't go trough code to verify it).

**Comment by ecdsa** [comment]
@sup2X thanks for the hint, will investigate that

**Comment by venzen** [comment]
In the end I worked around this issue by isolating the address that was giving Electrum grief.

The address had previously been used as change address by other wallets, but one particular address (that it received change from in the past) had not been imported to Electrum. Due to privacy concerns I did not want to import this particular address. Since the offending change address contained only a dust amount, I omitted it from a sweep of all my other addresses. Success and my immediate problem solved.

It was IMO that missing address (and related utxo that had dumped change into the offending address) that had caused the error described above.
